pest/src/iterators/pairs.rs

```diff
@@ -25,6 +25,82 @@ use super::queueable_token::QueueableToken;
 use super::tokens::{self, Tokens};
 use crate::RuleType;
 
+#[derive(Clone)]
+pub struct Cursor {
+    pub line: usize,
+    pub col: usize,
+}
+
+impl Default for Cursor {
+    fn default() -> Cursor {
+        Cursor { line: 1, col: 1 }
+    }
+}
+
+impl Cursor {
+    fn get(&self) -> (usize, usize) {
+        (self.line, self.col)
+    }
+}
+
+pub trait CursorPairs {
+    fn cursor(&self) -> Cursor;
+    fn cursor_mut(&mut self) -> &mut Cursor;
+
+    /// Move the (line, col) with string part
+    fn move_cursor(&mut self, part: &str) -> (usize, usize) {
+        let (l, c, has_new_line) = self.line_col(part);
+
+        let (prev_line, prev_col) = self.cursor().get();
+
+        self.cursor_mut().line += l;
+        if has_new_line {
+            self.cursor_mut().col = c;
+        } else {
+            self.cursor_mut().col += c;
+        }
+        (prev_line, prev_col)
+    }
+
+    /// Calculate line and col number of a string part
+    /// Fork from Pest for just count the part.
+    ///
+    /// https://github.com/pest-parser/pest/blob/85b18aae23cc7b266c0b5252f9f74b7ab0000795/pest/src/position.rs#L135
+    fn line_col(&self, part: &str) -> (usize, usize, bool) {
+        let mut chars = part.chars().peekable();
+
+        let mut line_col = (0, 0);
+        let mut has_new_line = false;
+
+        loop {
+            match chars.next() {
+                Some('\r') => {
+                    if let Some(&'\n') = chars.peek() {
+                        chars.next();
+
+                        line_col = (line_col.0 + 1, 1);
```

Comment:
Are these supposed to be `(line_col.0 + 1, line_col.1)`?

Reply:
This method has copied from AutoCorrect it has passed a lot of test case and use in production:

- https://github.com/huacnlee/autocorrect/blob/v2.5.5/autocorrect/src/result.rs#L287
- https://github.com/pest-parser/pest/blob/85b18aae23cc7b266c0b5252f9f74b7ab0000795/pest/src/position.rs#L135

And this line means:

```
If next char is `\n`, returns line + 1 and col reset to 1
```

Reply:
Ah yes, I see how that works now. Thanks.